django/contrib/gis/db/backends/postgis/base.py

```diff
@@ -1,17 +1,126 @@
+from functools import lru_cache
```

Comment:
This whole file still feels pretty much crazy, any ideas on how to do all those adapter types nicer?

Reply:
Okay, I think this code is now as good as it gets :D Still pretty bad but I guess it won't get much better.